**Universidad Internacional de La Rioja (UNIR) - Máster Universitario en Inteligencia Artificial - Procesamiento del Lenguaje Natural** 

***
Datos del alumno (Nombre y Apellidos):

Fecha:
***

<span style="font-size: 20pt; font-weight: bold; color: #0098cd;">Trabajo: Named-Entity Recognition</span>

**Objetivos** 

Con esta actividad se tratará de que el alumno se familiarice con el manejo de la librería spacy, así como con los conceptos básicos de manejo de las técnicas NER

**Descripción**

En esta actividad debes procesar de forma automática un texto en lenguaje natural para detectar características básicas en el mismo, y para identificar y etiquetar las ocurrencias de conceptos como localización, moneda, empresas, etc.

En la primera parte del ejercicio se proporciona un código fuente a través del cual se lee un archivo de texto y se realiza un preprocesado del mismo. En esta parte el alumno tan sólo debe ejecutar y entender el código proporcionado.

En la segunda parte del ejercicio se plantean una serie de preguntas que deben ser respondidas por el alumno. Cada pregunta deberá responderse con un fragmento de código fuente que esté acompañado de la explicación correspondiente. Para elaborar el código solicitado, el alumno deberá visitar la documentación de la librería spacy, cuyos enlaces se proporcionarán donde corresponda.

# Parte 1: carga y preprocesamiento del texto a analizar

Observa las diferentes librerías que se están importando.

In [54]:
import pathlib
import spacy
import pandas as pd
from spacy import displacy
import csv
import es_core_news_md

El siguiente código simplemente carga y preprocesa el texto. Para ello, lo primero que hace es cargar un modelo de lenguaje previamente entrenado. En este caso, se utiliza <i>es_core_news_md</i>: 

https://spacy.io/models/es#es_core_news_md


In [55]:
nlp = es_core_news_md.load()

El objeto <i>nlp</i> permite utilizar el modelo de lenguaje cargado, de forma que se puede procesar un texto y obtenerlo en su versión preprocesada. Así, nos permite realizar las diferentes tareas. En este caso, vamos a utilizar el pipeline para hacer un preprocesamiento básico, que consiste en tokenizar el texto.

In [56]:
filename = "./comentariosOdio.csv"

In [57]:

lines_number = 20
data = pd.read_csv(filename, delimiter=';', encoding='latin1')  
#data = pd.read_csv(filename, delimiter=';',nrows=lines_number)  

C:\Users\ccmarc4\AppData\Local\Temp\ipykernel_10272\916618882.py:2: DtypeWarning: Columns (5,6,7,8,9,10,11,12,13,14) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv(filename, delimiter=';', encoding='latin1')


El código anterior carga el archivo CSV (opcionalmente con un límite de líneas a leer) y genera la variable <i>data</i>, que contiene un Dataframe (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) con los datos leídos del CSV.

Te vendrá bien conocer la siguiente documentación:
<ul>
    <li>https://spacy.io/api/doc</li>
    <li>https://spacy.io/api/token</li>
    <li>https://spacy.io/api/morphology#morphanalysis</li>
</ul>

### Playground

Utiliza este espacio para hacer pruebas y ensayos con las variables generadas con el código previo. A modo de ejemplo, se ofrece código que realiza las siguientes tareas: 


- leer un número dado de líneas del Dataframe y generar dos listas con los valores (se pueden leer directamente del DataFrame, se muestra el ejemplo como una opción más)
- procesar el texto de cada comentario


Para procesarlo, hay utilizar el objeto <i>nlp</i> y así obtener objetos de la clase <i>Doc</i> (https://spacy.io/api/doc)

Visita la documentación de dicha clase y experimenta probando las diferentes funciones y atributos 

In [58]:
# Puedes insertar aquí código de pruebas para experimentar con las diferentes funciones y atributos de 'doc'.
#print(data["CONTENIDO A ANALIZAR"][1])
#print(data["INTENSIDAD"][1])
doc = []
value = []

#con el bucle, generamos sendas listas con los comentarios ya parseados y con el valor de intensidad
for i in range(0, lines_number):#'''len(data["CONTENIDO A ANALIZAR"])'''
    
    #en un primer paso se parsea el comentario. En el segundo paso se añade el objeto a la lista
    tmp_doc = nlp(data["CONTENIDO A ANALIZAR"][i])
    doc.append(tmp_doc)
    
    #en un primer paso extrae el valor. En el segundo paso se añade el valor a la lista
    tmp_value = data["INTENSIDAD"][i]
    value.append(tmp_value)


# #ejemplo de cómo recorrer un comentario palabra por palabra    
# for token in doc[1]:
#     print(token)

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 1.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántos registros contiene el corpus?</span>

He visto que hay varios problemas en el dataset:
- Muchas filas en las que todos los valores son nulos menos 1, que es la continuación del contenido de la fila anterior. Para arreglar esto borraré las filas que solo contengan ese valor.
- Hay varias columnas más de las que son en verdad. Esto se debe a que hay direcciones, nombres y demás mal tabulados. Borraré todas las columnas que no tengan un nombre definido, ya que estas columnas en realidad no tienen nombre.

In [59]:
import pandas as pd

# Configura el nombre de tu archivo CSV original y el de salida
input_file = "comentariosOdio.csv"
output_file = "comentariosOdioLimpio.csv"

# Lee el archivo con utf-8 y evita errores de codificación
try:
    with open(input_file, encoding='utf-8', errors='replace') as file:
        df = pd.read_csv(file, delimiter=';', engine='python')
except Exception as e:
    print(f"Error al leer el archivo: {e}")
    exit()

# Guarda el número de columnas base (número de columnas en la primera fila)
num_columnas_base = 9

# Encuentra columnas con índice mayor al número de columnas base
exceso_columnas = df.columns[num_columnas_base:]

# Lista para almacenar las filas con datos en columnas excedentes
filas_con_datos_excedentes = []

# Verifica si alguna de las columnas excedentes tiene datos
for col in exceso_columnas:
    if df[col].notnull().any():
        # Obtén las filas donde esta columna tiene datos
        filas_con_datos = df[df[col].notnull()]
        filas_con_datos_excedentes.append(filas_con_datos)
        # Imprime cada fila que tiene datos en una columna excedente
        for idx, fila in filas_con_datos.iterrows():
            print(f"Fila con datos en columna excedente ({col}):")
            print(fila)
            print("-" * 40)

# Elimina las columnas excedentes del DataFrame
df = df.iloc[:, :num_columnas_base]

# Imprime un resumen
print(f"Columnas eliminadas: {len(exceso_columnas)}")
print(f"Filas afectadas por columnas eliminadas: {sum([len(f) for f in filas_con_datos_excedentes])}")

Fila con datos en columna excedente (Unnamed: 9):
MEDIO                                                           alcoi, 66
SOPORTE                                                    en mislata, 62
URL                                                           en elda, 61
TIPO DE MENSAJE                                              en xàbia, 59
CONTENIDO A ANALIZAR                                      el campello, 53
INTENSIDAD                                             la vila joiosa, 52
TIPO DE ODIO                                     burjassot y l'eliana, 51
TONO HUMORISTICO                                              godella, 49
MODIFICADOR                                                   manises, 48
Unnamed: 9                                                    alfafar, 47
Unnamed: 10                                                     puçol, 46
Unnamed: 11                                                benicàssim, 45
Unnamed: 12                               benicarló, petrer y 

In [60]:
#limpiar el df quitando todas las columnas menos CONTENIDO A ANALIZAR e INTENSIDAD
df = df[["CONTENIDO A ANALIZAR", "INTENSIDAD"]]

In [61]:
# Si la fila es nula o duplicada, eliminar
df = df.dropna()
df = df.drop_duplicates()

# contar cuantas filas tiene el df
print(f"Filas en el DataFrame: {len(df)}")

Filas en el DataFrame: 485712


In [62]:
# Guarda el archivo limpio
try:
    df.to_csv(output_file, sep=';', index=False, encoding='utf-8')
    print(f"Archivo limpio guardado como '{output_file}'")
except Exception as e:
    print(f"Error al guardar el archivo: {e}")

Archivo limpio guardado como 'comentariosOdioLimpio.csv'


In [63]:
# Buscar filas nulas o repetidas

# Crea una lista de filas con valores nulos
filas_nulas = df[df.isnull().any(axis=1)]

# Crea una lista de filas duplicadas
filas_duplicadas = df[df.duplicated()]

# Imprime un resumen
print(f"Columnas con valores nulos: {filas_nulas}")
print(f"Filas duplicadas: {len(filas_duplicadas)}")

Columnas con valores nulos: Empty DataFrame
Columns: [CONTENIDO A ANALIZAR, INTENSIDAD]
Index: []
Filas duplicadas: 0


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>

Después de limpiar el dataset de columnas innecesarias, filas nulas y valores duplicados, el dataset contiene 485712 registros si miramos el número de filas del CSV

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 2.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántas palabras totales hay en los comentarios del corpus?</span>

In [64]:
data = pd.read_csv("comentariosOdioLimpio.csv", delimiter=';', encoding= 'utf-8')

if "CONTENIDO A ANALIZAR" in data.columns:
    # Combinar todos los comentarios de la columna en un solo texto
    texto_completo = " ".join(data["CONTENIDO A ANALIZAR"].astype(str))

    # Contar las palabras
    numero_palabras = len(texto_completo.split())

    print(f"Número total de palabras en el corpus: {numero_palabras}")
else:
    print("La columna 'CONTENIDO A ANALIZAR' no existe en el dataset.")

Número total de palabras en el corpus: 60518520


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>

Después de eliminar las columnas repetidas, nulas y las columnas que no son relevantes, si revisamos el contenido de la columna de comentarios llamada "CONTENIDO A ANALIZAR", podemos observar que el corpus contiene un total de 60518520 palabras.


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 3.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuál el número promedio de palabras en cada comentario?</span>

In [65]:
def contar_palabras(texto):
    if isinstance(texto, str):
        return len(texto.split())
    return 0

In [66]:
columna = "CONTENIDO A ANALIZAR"
if columna in data.columns:

    comentarios = data[columna]
    
    # Aplicamos la función a cada comentario
    data["num_palabras"] = comentarios.apply(contar_palabras)
    
    # Calcular el promedio de palabras por comentario
    promedio = data["num_palabras"].mean()
    
    print(f"Número promedio de palabras por comentario: {promedio}")
    
    print("\nEstadísticas adicionales:")
    print(data["num_palabras"].describe())
else:
    print(f"La columna '{columna}' no existe en el dataset.")

Número promedio de palabras por comentario: 124.5975392825378

Estadísticas adicionales:
count    485712.000000
mean        124.597539
std         272.501166
min           0.000000
25%          10.000000
50%          19.000000
75%          59.000000
max        5341.000000
Name: num_palabras, dtype: float64


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>

El número medio de palabras por comentario es de aproximadamente  124.6. 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 4.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál el número promedio de palabras en los comentarios de cada grupo?</span>

In [67]:
#implrimir 20 valores de la comlumna INTENSIDAD y su type
print(data["INTENSIDAD"].head(20))
print(data["INTENSIDAD"].dtype)

# Convertir la columna INTENSIDAD a tipo numérico
data["INTENSIDAD"] = pd.to_numeric(data["INTENSIDAD"], errors='coerce')

# Verificar si la conversión fue exitosa
print(data["INTENSIDAD"].dtype)

0     3.0
1     0.0
2     3.0
3     3.0
4     3.0
5     3.0
6     4.0
7     3.0
8     4.0
9     3.0
10    3.0
11    4.0
12    0.0
13    4.0
14    4.0
15    3.0
16    0.0
17    3.0
18    0.0
19    0.0
Name: INTENSIDAD, dtype: object
object
float64


In [68]:
# media de palabras por comentario si intensidad == 0 y para intensidad >0

# Filtrar comentarios con intensidad 0
comentarios_no_odio = data[data["INTENSIDAD"] == 0]

# Filtrar comentarios con intensidad mayor a 0
comentarios_odio = data[data["INTENSIDAD"] > 0]

# Calcular el promedio de palabras por comentario
promedio_no_odio = comentarios_no_odio["num_palabras"].mean()
promedio_odio = comentarios_odio["num_palabras"].mean()

print(f"Número promedio de palabras por comentario (sin odio): {promedio_no_odio}")
print(f"Número promedio de palabras por comentario (con odio): {promedio_odio}")


Número promedio de palabras por comentario (sin odio): 127.16041126120524
Número promedio de palabras por comentario (con odio): 16.69876630868909


In [69]:
#borrar columna
data = data.drop(columns=["num_palabras"])

<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
 
Si filtramos los datos de filas donde no se dice si hay odio y en las que si hay odio y luego hacemos la media, podemos ver que la media de palabras en comentarios de odio es de 16.69. 

Mientras que la media de palabras en comentarios que no son de odio es de 127.16.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 5.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el número promedio de oraciones en los comentarios de cada grupo?</span>

In [70]:
def contar_oraciones(texto):
    if isinstance(texto, str):
        # Dividir por ".", "?" o "!" y eliminar vacíos
        return len([oracion.strip() for oracion in texto.split('.') if oracion.strip()])
    return 0

In [71]:
# contar nº de oraciones en cada comentario de odio o no odio

# Filtrar comentarios con intensidad 0
comentarios_no_odio = data[data["INTENSIDAD"] == 0]

# Filtrar comentarios con intensidad mayor a 0
comentarios_odio = data[data["INTENSIDAD"] > 0]

# Aplicar la función a cada comentario
comentarios_no_odio["num_oraciones"] = comentarios_no_odio["CONTENIDO A ANALIZAR"].apply(contar_oraciones)
comentarios_odio["num_oraciones"] = comentarios_odio["CONTENIDO A ANALIZAR"].apply(contar_oraciones)

# Calcular el promedio de oraciones por comentario
promedio_no_odio = comentarios_no_odio["num_oraciones"].mean()
promedio_odio = comentarios_odio["num_oraciones"].mean()

print(f"Número promedio de oraciones por comentario (sin odio): {promedio_no_odio}")
print(f"Número promedio de oraciones por comentario (con odio): {promedio_odio}")

Número promedio de oraciones por comentario (sin odio): 5.864360945361192
Número promedio de oraciones por comentario (con odio): 1.6069938759208307


C:\Users\ccmarc4\AppData\Local\Temp\ipykernel_10272\1189685917.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  comentarios_no_odio["num_oraciones"] = comentarios_no_odio["CONTENIDO A ANALIZAR"].apply(contar_oraciones)
C:\Users\ccmarc4\AppData\Local\Temp\ipykernel_10272\1189685917.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  comentarios_odio["num_oraciones"] = comentarios_odio["CONTENIDO A ANALIZAR"].apply(contar_oraciones)


In [72]:
data

CONTENIDO A ANALIZAR  INTENSIDAD
0       el barça nunca acaeza ante un segundo b ni ant...         3.0
1       el real madrid ha puesto punto y final a su an...         0.0
2       cristina cifuentes podría haber sido la presid...         3.0
3       habría que reabrir el caso. el supremo se dedi...         3.0
4       me parece un poco exagerado pedir más de tres ...         3.0
...                                                   ...         ...
485707  que mala es ayuso , que cobra menos impuestos ...         4.0
485708  quién paga la factura de las "televisiones" de...         1.0
485709  sabéis lo que está abultada?\nel tamaño de nue...         4.0
485710  sólo queréis el dinero público para que se inv...         4.0
485711  un panfleto, criticando el gasto públco????soi...         4.0

[485712 rows x 2 columns]

<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
 
Las oraciones se suelen separar por ".", "?" o "!", por lo que si separamos las oraciones por ahí y hacemos la media de oraciones por cada grupo, nos da que el promedio de oraciones en comentarios de odio es de aproximadamente 1.61.

Mientras que el promedio de oraciones en comentarios que no son de odio es de aproximadamente 5.86.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 6.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el porcentaje de comentarios que contienen entidades NER en cada grupo?</span>

A partir de ahora, tras muchos experimentos donde me daba un error en el kernel, voy a hacer las cosas con listas, de esta forma voy a probar que no se reinicie. 

In [ ]:
batch_size = 50  # Ajusta según la capacidad de tu máquina

doc = []
value = []

for start in range(0, 4999, batch_size):
    end = start + batch_size
    batch_texts = data["CONTENIDO A ANALIZAR"][start:end]
    
    # Procesar en lote con SpaCy para optimización
    docs_batch = list(nlp.pipe(batch_texts))  # Usa pipe para procesar por lotes
    doc.extend(docs_batch)
    value.extend(data["INTENSIDAD"][start:end])

print(doc[0])
print(value[0])

el barça nunca acaeza ante un segundo b ni ante un tercera , ya estan los arbitros para impedirlo....lo de messi es una autentica vergüenza
3.0


In [ ]:
doc[0].ents

(messi,)

In [ ]:
len(doc[0].ents)

1

In [ ]:
# Inicializar contadores
total_odio = 0
total_no_odio = 0
con_entidades_odio = 0
con_entidades_no_odio = 0

# Recorrer la lista de documentos y contar entidades NER
for i in range(len(doc)):
    if value[i] > 0:
        total_odio += 1
        if len(doc[i].ents) > 0:
            con_entidades_odio += 1
    else:
        total_no_odio += 1
        if len(doc[i].ents) > 0:
            con_entidades_no_odio += 1

# Calcular porcentajes
porcentaje_odio_con_entidades = (con_entidades_odio / total_odio) * 100 if total_odio > 0 else 0
porcentaje_no_odio_con_entidades = (con_entidades_no_odio / total_no_odio) * 100 if total_no_odio > 0 else 0

print(f"Porcentaje de comentarios de odio con entidades NER: {porcentaje_odio_con_entidades:.2f}%")
print(f"Porcentaje de comentarios no de odio con entidades NER: {porcentaje_no_odio_con_entidades:.2f}%")


Porcentaje de comentarios de odio con entidades NER: 39.57%
Porcentaje de comentarios no de odio con entidades NER: 38.85%


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
 
Tras guardar en una lista cada comentario y en otra cada una de las valoraciones de odio, podemos calcular el porcentaje de coementarios de odio y no odio con entidades NER, el cual es este:

Porcentaje de comentarios de odio con entidades NER: 36.50%
Porcentaje de comentarios no de odio con entidades NER: 59.37%

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 7.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el porcentaje de comentarios que contienen entidades NER de tipo PERSON en cada grupo?</span>

In [ ]:
print(doc[1].ents)
print(len(doc[1].ents))
print(doc[1].ents[2].label_)

(real madrid, copa del rey, zidane, madrid, zidane)
5
PER


In [ ]:
# # Inicializar contadores
# total_odio = 0
# total_no_odio = 0
# con_entidades_person_odio = 0
# con_entidades_person_no_odio = 0

# # Recorrer la lista de documentos y contar entidades NER de tipo PERSON
# for i in range(len(doc)):
#     texto_procesado = nlp(doc[i])  # Procesar el texto con el modelo de NLP
    
#     if value[i] > 0:  # Comentario de odio
#         total_odio += 1
#         if any(ent.label_ == "PER" for ent in texto_procesado.ents):  # Comprobar si hay una entidad de tipo PERSON
#             con_entidades_person_odio += 1
#     else:  # Comentario no de odio
#         total_no_odio += 1
#         if any(ent.label_ == "PER" for ent in texto_procesado.ents):  # Comprobar si hay una entidad de tipo PERSON
#             con_entidades_person_no_odio += 1

# # Calcular porcentajes
# porcentaje_odio_con_entidades_person = (con_entidades_person_odio / total_odio) * 100 if total_odio > 0 else 0
# porcentaje_no_odio_con_entidades_person = (con_entidades_person_no_odio / total_no_odio) * 100 if total_no_odio > 0 else 0

# # Imprimir resultados
# print(f"Porcentaje de comentarios de odio con entidades NER de tipo PERSON: {porcentaje_odio_con_entidades_person:.2f}%")
# print(f"Porcentaje de comentarios no de odio con entidades NER de tipo PERSON: {porcentaje_no_odio_con_entidades_person:.2f}%")

In [ ]:
# # Frase para analizar
# texto = "Lionel Messi juega en el Inter Miami y ganó el Balón de Oro en 2023."

# # Procesar el texto con el modelo de SpaCy
# doc1 = nlp(texto)

# # Identificar y contar las entidades
# entidades = [(ent.text, ent.label_) for ent in doc1.ents]  # Lista de entidades con sus etiquetas
# total_entidades = len(entidades)  # Contar cuántas entidades hay

# # Mostrar los resultados
# print("Entidades encontradas:")
# for entidad, etiqueta in entidades:
#     print(f"Texto: {entidad}, Etiqueta: {etiqueta}")

# print(f"Total de entidades: {total_entidades}")

<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 8.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el porcentaje de palabras en cada combinación posible de género y número (p.ej. masculino singular) en cada grupo?</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 9.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio), indica cuántas entidades de cada tipo posible se reconocen en cada uno de los grupos</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 10.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio), extrae y muestra los 100 lemas más repetidos en los comentarios de cada grupo</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 11.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Es posible utilizar alguna de las características extraídas en las preguntas anteriores para determinar si un mensaje contiene odio? Justifica tu respuesta con el análisis estadístico que consideres necesario.</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
 